rollup/rollupsyncservice/rollupsyncservice.go

```diff
@@ -0,0 +1,395 @@
+	for from := s.latestProcessedBlock + 1; from <= latestConfirmed; from += defaultFetchBlockRange {
+		if s.ctx.Err() != nil {
+			return
+		}
+
+		to := from + defaultFetchBlockRange - 1
+		if to > latestConfirmed {
+			to = latestConfirmed
+		}
+
+		logs, err := s.client.fetchRollupEventsInRange(s.ctx, from, to)
+		if err != nil {
+			log.Error("failed to fetch rollup events in range", "from block", from, "to block", to, "err", err)
+			return
+		}
+
+		if err := s.parseAndUpdateRollupEventLogs(logs, to); err != nil {
+			log.Error("failed to parse and update rollup event logs", "err", err)
+			return
+		}
+	}
+}
+
+func (s *RollupSyncService) parseAndUpdateRollupEventLogs(logs []types.Log, endBlockNumber uint64) error {
+	for _, vLog := range logs {
+		switch vLog.Topics[0] {
+		case s.l1CommitBatchEventSignature:
+			event := &L1CommitBatchEvent{}
+			if err := UnpackLog(s.scrollChainABI, event, "CommitBatch", vLog); err != nil {
+				return fmt.Errorf("failed to unpack commit rollup event log, err: %w", err)
+			}
+			batchIndex := event.BatchIndex.Uint64()
+
+			chunkBlockRanges, err := s.getChunkRanges(batchIndex, &vLog)
+			if err != nil {
+				return fmt.Errorf("failed to get chunk ranges, err: %w", err)
+			}
+			rawdb.WriteBatchChunkRanges(s.db, batchIndex, chunkBlockRanges)
+
+		case s.l1RevertBatchEventSignature:
+			event := &L1RevertBatchEvent{}
+			if err := UnpackLog(s.scrollChainABI, event, "RevertBatch", vLog); err != nil {
+				return fmt.Errorf("failed to unpack revert rollup event log, err: %w", err)
+			}
+			batchIndex := event.BatchIndex.Uint64()
+
+			rawdb.DeleteBatchChunkRanges(s.db, batchIndex)
+
+		case s.l1FinalizeBatchEventSignature:
+			event := &L1FinalizeBatchEvent{}
+			if err := UnpackLog(s.scrollChainABI, event, "FinalizeBatch", vLog); err != nil {
+				return fmt.Errorf("failed to unpack finalized rollup event log, err: %w", err)
+			}
+			batchIndex := event.BatchIndex.Uint64()
+
+			parentBatchMeta, chunks, err := s.getLocalInfoForBatch(batchIndex)
+			if err != nil {
+				return fmt.Errorf("failed to get local node info, batch index: %v, err: %w", batchIndex, err)
+			}
+
+			if err := validateBatch(event, parentBatchMeta, chunks, s.node); err != nil {
+				return fmt.Errorf("fatal: validateBatch failed: finalize event: %v, err: %w", *event, err)
+			}
+			endChunk := chunks[len(chunks)-1]
+			endBlock := endChunk.Blocks[len(endChunk.Blocks)-1]
+
+			rawdb.WriteFinalizedL2BlockNumber(s.db, endBlock.Header.Number.Uint64())
+			rawdb.WriteFinalizedBatchMeta(s.db, batchIndex, calculateFinalizedBatchMeta(parentBatchMeta, event.BatchHash, chunks))
+
+			if batchIndex%100 == 0 {
+				log.Info("finalized batch progress", "batch index", batchIndex, "finalized l2 block height", endBlock.Header.Number.Uint64())
+			}
+
+		default:
+			return fmt.Errorf("unknown event, topic: %v, tx hash: %v", vLog.Topics[0].Hex(), vLog.TxHash.Hex())
+		}
+	}
+
+	// note: the batch updates above are idempotent, if we crash
```

Comment:
Btw, `DeleteBatchChunkRanges` will succeed even if the key does not exist, right?

Reply:
nice catch.

Reply:
Yes. ref: https://pkg.go.dev/github.com/syndtr/goleveldb/leveldb#DB.Delete
also adding a test to ensure the delete operation handles non-existing values without errors (but only in the memory database, not leveldb).